auth/src/main/java/org/aerogear/mobile/security/AsyncSecurityCheckExecutor.java

```diff
@@ -101,16 +105,37 @@ public AsyncSecurityCheckExecutor build() {
         final MetricsService metricsService = getMetricsService();
         final Map<String, Future<SecurityCheckResult>> res = new HashMap<>();
 
-        for (final SecurityCheck check : getChecks()) {
-            res.put(check.getName(), (executorService.submit(() -> {
-                final SecurityCheckResult result =  check.test(getContext());
-                if (metricsService != null) {
-                    metricsService.publish(new SecurityCheckResultMetric(result));
-                }
+        final Collection<SecurityCheck> checks = getChecks();
```

Comment:
Wouldn't be easier like this:
```java
public Map<String, Future<SecurityCheckResult>> execute() {

        final MetricsService metricsService = getMetricsService();
        final Map<String, Future<SecurityCheckResult>> res = new HashMap<>();

        final Set<SecurityCheckResult> results = new HashSet<>();
        final Collection<SecurityCheck> checks = getChecks();

        for (final SecurityCheck check : checks) {
            res.put(check.getName(), (executorService.submit(() -> {
                final SecurityCheckResult result =  check.test(getContext());
                results.add(result);

                synchronized (results) {
                    if (metricsService != null && results.size() == checks.size()) {
                        metricsService.publish(new SecurityCheckResultMetric(result));
                    }
                }

                return result;
            })));
        }

        return res;
    }
```

Reply:
I see, wasn't familiar with synchronized methods and blocks.

I was, however, thinking about ways that users could do this kind of queuing of actions themselves. Both of these approaches are only possible because we're in control of the Callables or aware of the ExecutorService.

I guess we can postpone these improvements for Rx :pray:

Reply:
@paolobueno TBH I think we shouldn't change the executor to batch the metrics, but batching should be done by the metric service (see my comment at the end of the PR)